src/main/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandler.java

```diff
@@ -99,32 +117,68 @@ public PaymentHandleResponse createPayment(@Nonnull String ctpPaymentId) {
 
                                 Payment paypalPlusPayment = paymentMapper.ctpPaymentToPaypalPlus(
                                         new CtpPaymentWithCart(optPayment.get(), optCart.get()));
-
-                                return paypalPlusFacade.getPaymentService().create(paypalPlusPayment)
-                                        .thenCompose(createdPpPayment -> setApprovalUrlAndInterfaceId(createdPpPayment, ctpPaymentId))
+                                AddInterfaceInteraction action = createAddInterfaceInteractionAction(paypalPlusPayment, REQUEST);
+                                // 1. save create paypal request to ctp payment's interface interaction
+                                return ctpFacade.getPaymentService().updatePayment(ctpPayment, Collections.singletonList(action))
+                                        .thenCompose(payment -> {
+                                            // 2. create paypal payment
+                                            return paypalPlusFacade.getPaymentService().create(paypalPlusPayment)
+                                                    .thenCompose(createdPpPayment -> {
+                                                                List<UpdateAction<io.sphere.sdk.payments.Payment>> updateActions
+                                                                        = new ArrayList<>(getApprovalUrlAndInterfaceIdAction(createdPpPayment));
+                                                                AddInterfaceInteraction interaction = createAddInterfaceInteractionAction(createdPpPayment, RESPONSE);
+                                                                updateActions.add(interaction);
+                                                                // 3. save approval url and payment ID and response from paypal to ctp payment
+                                                                return ctpFacade.getPaymentService().updatePayment(payment, updateActions)
```

Comment:
i'd carry out this part out of this completion stage (save approval url and payment ID and response from paypal to ctp payment) one level up: it is about CTP object update, but you are doing in this stage papal plus payment creation.

In another words: don't call _ctpFacade_ services inside _paypalPlusFacade_ services chain.

Reply:
But I need both CTP payment and Paypal Plus payment.

Reply:
In _Intellij IDEA_: `Cmd+Alt+M` -> _Extract Method_:

```java
................................
return ctpFacade.getPaymentService().updatePayment(ctpPayment, Collections.singletonList(action))                                        
.thenCompose(getPaymentCompletionStageFunction(paypalPlusPayment))
.thenApply(PaymentMapper::getApprovalUrl)
.thenApply(approvalUrlOpt -> approvalUrlOpt
.............

// Just give more reasonable function name
private Function<io.sphere.sdk.payments.Payment, CompletionStage<Payment>> getPaymentCompletionStageFunction(Payment paypalPlusPayment) {
    return payment -> {
        // 2. create paypal payment
        return paypalPlusFacade.getPaymentService().create(paypalPlusPayment)
                .thenCompose(createdPpPayment -> {
                            List<UpdateAction<io.sphere.sdk.payments.Payment>> updateActions
                                    = new ArrayList<>(getApprovalUrlAndInterfaceIdAction(createdPpPayment));
                            AddInterfaceInteraction interaction = createAddInterfaceInteractionAction(createdPpPayment, RESPONSE);
                            updateActions.add(interaction);
                            // 3. save approval url and payment ID and response from paypal to ctp payment
                            return ctpFacade.getPaymentService().updatePayment(payment, updateActions)
                                    .thenApply(ignore -> createdPpPayment);
                        }
                );
    };
}
```

Reply:
@andrii-kovalenko-ct @lojzatran yes, this is exactly what I meant.

Reply:
https://github.com/commercetools/commercetools-paypal-plus-integration/pull/49/commits/191dad9b40d6ab6ebdd4392d864f93a07e45a8c9

Reply:
https://github.com/commercetools/commercetools-paypal-plus-integration/pull/49/commits/8b8630fd3fa0bedf3778d831871058aff3734c62

Reply:
thx, now my eyes are not bleeding any more :)